### ISSUE TITLE
Location hierarchy formatting issues

### ISSUE THREAD

**Issue by kairstenfay** [nextstrain/ncov-ingest#78]
The [source-data/location_hierarchy.tsv](https://github.com/nextstrain/ncov-ingest/blob/master/source-data/location_hierarchy.tsv) file contains some issues where 
* some rows only have 3 columns instead of 4 (which GitHub currently points out when you view this file in the browser), i.e. a missing final tab
* some rows have 5 columns instead of 4, i.e. a surplus final tab

To prevent future formatting issues with the manually updated location hierarchy file, I suggest adding a script like `bin/clean-location-hierachy` that either
* does some bash magic (something like `cut -f1-4 source-data/location-hierarchy.tsv` to grab only the first four columns, then something else I haven't quite figured out to add a trailing tab to rows missing the 4th column -- awk and gsub()?), or
* loads it into pandas and print it back out again with only 4 columns, tab-delimited.

I also suggest sorting the file on region, country, division, and location in case a row is misplaced. 

This cleaning script should be added to the ingest scripts.

**Comment by joverlee521** [comment]
Just had a thought maybe we should make this a new workflow to run whenever there's a push to `source-data/location_hierarchy.tsv` instead of within the ingest scripts.

**Comment by kairstenfay** [comment]
> Just had a thought maybe we should make this a new workflow to run whenever there's a push to `source-data/location_hierarchy.tsv` instead of within the ingest scripts.

That's a great idea!

**Comment by eharkins** [comment]
We could also add an option to the script to add in the the output from [the slack location hierarchy additions file](https://github.com/nextstrain/ncov-ingest/blob/eca007ed26a8832c0ebfcc47d43cb016e8776f26/bin/notify-on-location-hierarchy-addition) so that you could run it locally to add those new location hierarchies and then manually remove the wrong ones or duplicates (there will be fewer duplicates with this clean up script included in automation) instead of manually adding the correct new ones. 

Sorry if this belongs better in https://github.com/nextstrain/ncov/pull/455

**Comment by kairstenfay** [comment]
> We could also add an option to the script to add in the the output from [the slack location hierarchy additions file](https://github.com/nextstrain/ncov-ingest/blob/eca007ed26a8832c0ebfcc47d43cb016e8776f26/bin/notify-on-location-hierarchy-addition) so that you could run it locally to add those new location hierarchies and then manually remove the wrong ones or duplicates (there will be fewer duplicates with this clean up script included in automation) instead of manually adding the correct new ones.
> 
> Sorry if this belongs better in [nextstrain/ncov#455](https://github.com/nextstrain/ncov/pull/455)

So if I understand correctly, you're suggesting an automated way of appending the new location hierarchies to _source-data/location_hierarchy.tsv_? Thus saving copy/pasting by build maintainers? 

In general, I'm very interested in learning of more ways to improve the nCoV build maintenance workflow. IMO such a feature doesn't thematically fit within a "cleaning" script, but it could influence the overall solution that resolves this issue since sorting would have to occur after appending to the location hierarchy. So, thanks for bringing it up here! I haven't looked at nextstrain/ncov#455, but I think this repo, perhaps under a new issue, is a fine place for this discussion.

**Comment by eharkins** [comment]
> So if I understand correctly, you're suggesting an automated way of appending the new location hierarchies to source-data/location_hierarchy.tsv? Thus saving copy/pasting by build maintainers?

@kairstenfay exactly, and perhaps this is not top on the list in terms of potential time saved by automating things currently achieved by copy/pasting. I'll open a new issue anyway. Thanks!

**Comment by kairstenfay** [comment]
> @kairstenfay exactly, and perhaps this is not top on the list in terms of potential time saved by automating things currently achieved by copy/pasting. I'll open a new issue anyway. Thanks!

Thanks for taking the initiative on that. Perhaps the feature you described could fit in the new location hierarchy workflow @joverlee521 proposed.

**Comment by kairstenfay** [comment]
> > @kairstenfay exactly, and perhaps this is not top on the list in terms of potential time saved by automating things currently achieved by copy/pasting. I'll open a new issue anyway. Thanks!
> 
> Thanks for taking the initiative on that. Perhaps the feature you described could fit in the new location hierarchy workflow @joverlee521 proposed.

I think naming the new script and/or workflow "adjust location hierarchy" instead of "clean location hierarchy" would create a logical place for @eharkins's suggested feature to live.

**Comment by emmahodcroft** [comment]
My only concern with the automation is that it's important to be checking these new locations are reasonable - that the division is a division and the location is a location (it's not uncommon to get cities as divisions, for example). It makes things harder down the line when we end up with too much inconsistency in how we're dividing things  and what things are called.... I'm all for making things easier, but I would want to be sure we keep doing these checks rather than allowing everything to be added automatically!

**Comment by eharkins** [comment]
I agree @emmahodcroft; I think it would make sense to clean tabs and sort location hierarchies in an automated workflow, but adding new ones should still be a manual process. Here's what I propose for adding new locations manually/locally:
1. Run `cat location_hierarchy_additions.tsv >> location_hierarchy.tsv` using the new location hierarchies from slack (this achieves what I was hoping for assuming you correct errors in the steps below)
2. Run the sorting and tab clean up script being discussed here on `location_hierarchy.tsv`
3. Manually check that no bad hierarchies were added to `location_hierarchy.tsv`, removing duplicates, mispellings, and ensure
> division is a division and the location is a location
4. Follow the normal protocol for adding the valid new locations to `ncov/defaults/{color_ordering.tsv, lat_longs.tsv}` and annotating invalid locations in `ncov-ingest/source-data/gistad_annotations.tsv`

**Comment by emmahodcroft** [comment]
I'd agree with this except I'd reverse the order - check the locations first in `location_hierarchy_additions.tsv` to ensure they're 'legit' and remove any that shouldn't be there - _then_ `cat`, sort, and check the remaining 'good' ones. Otherwise it can be difficult to remember what's new, and to be sure you scrolled around, found it, and removed it. 

Also important to remember do to this with the 'final' version of the file - after any changes to `gisaid_annotations.tsv`! Certainly for me yesterday this was key as the additions shrunk when I'd fixed up the sequences. This does reinforce being able to generate new `location_hierarchy_additions.tsv` files without necessarily re-running a build though!

**Comment by eharkins** [comment]
While we are autosorting things, are there other files, e.g. gisaid_annotations.tsv that would be good to run basic sorting (and perhaps tab cleanup) on as well on push?

**Comment by kairstenfay** [comment]
> While we are autosorting things, are there other files, e.g. gisaid_annotations.tsv that would be good to run basic sorting (and perhaps tab cleanup) on as well on push?

Excellent point! I am happy to also adjust the annotations files if that is desirable among our build maintainers. We may want to do a one-time cleanup of duplicates in annotations files, though, before sorting them. I would guess we would keep the entry listed lower in the file in the event of a duplicate.

**Comment by emmahodcroft** [comment]
Would also be happy to do this! 

One point is that we've kept the `title` and `paper_url` at the bottom half of the file just to make it easier for shepherds to add the other stuff at the top (Paul is exclusively working on adding papers, so this just keeps all that separate). So when sorting we would want to maintain that if possible. 

Alternatively we could split the paper stuff out to an entirely separate file - I suggested this a while ago and meant to do it, but it hasn't come to pass yet. Would be happy for either solution! (The two files could then essentially be `cat` together before being passed to the scripts that ingest them, as there's no real difference between the two except for humans!)

**Comment by eharkins** [comment]
> We may want to do a one-time cleanup of duplicates in annotations files, though, before sorting them. I would guess we would keep the entry listed lower in the file in the event of a duplicate.

@kairstenfay removing duplicates sounds good, just in case they cause confusion or take up space. If we are removing exact duplicate annotations (all columns exactly the same), does it matter which one we keep?

> One point is that we've kept the title and paper_url at the bottom half of the file just to make it easier for shepherds to add the other stuff at the top

@emmahodcroft if we were auto sorting them using a script on pushing to this repo, we could just stick any new annotations on the end of the file and they would get sorted into place by the script when we push, right? It still might be useful to be able to look for existing annotations in the file in addition to / in the process of appending new ones, but searching the file would be easier if it had been autosorted.

But if having the paper annotations potentially interleaved with the rest due to autosorting made life harder in a way I'm not getting, this seems like a good solution:
> Alternatively we could split the paper stuff out to an entirely separate file - I suggested this a while ago and meant to do it, but it hasn't come to pass yet. Would be happy for either solution! (The two files could then essentially be cat together before being passed to the scripts that ingest them, as there's no real difference between the two except for humans!)

**Comment by emmahodcroft** [comment]
It's not a huge deal, except that the `paper_url` and `title` info is about 2/3 of the file right now. And generally, it's info we wouldn't ever be interested in browsing or looking into. So, it's just that if it was mixed in, then it would just be *most* of the lines, which makes general browsing or looking for 'actually interesting' information harder. As I said, probably not a huge deal, but probably also not a big deal to either separate these out and sort them then just print one after another, or split into two files :)

**Comment by eharkins** [comment]
@kairstenfay would cleaning up and sorting annotations belong as an (optional) feature of this script https://github.com/nextstrain/ncov-ingest/blob/master/bin/check-annotations ?

As for location hierarchies it seems to make more sense to (as discussed above) create `bin/adjust-location-hierarchy` which runs upon pushing changes to `source-data/location_hierarchy.tsv`. Since https://github.com/nextstrain/ncov-ingest/blob/master/bin/check-locations seems like it creates the `location-hierarchies-additions.tsv` and is more geared toward the ingest workflow so it wouldn't be as good of a place for sorting and cleaning newly added locations. Sorry if that is obvious, just checking my interpretation makes sense as I orient myself more with the scripts in the repo.
